### ISSUE TITLE
Additional features required for production packaging

### ISSUE THREAD

**Issue by hendrenja** [cortoproject/corto#409]
This issue should open a dialogue for compiling list of requirements for packaging Corto project binaries. 
# Currently Supported
- Archive support via tar
- Architecture specification (x64, i386, etc)
- Versioning (Major-Minor-Release)
# Missing
- Manually specify version (Major-Minor-Release)
  - ./corto/version.txt ?
  - Use cases: hot-patches, bug-fixes, ?
- Debug, Development, and Release build support

**Comment by SanderMertens** [comment]
Currently Corto uses semantic versioning for it's core library. There is no support for package versioning yet. What we have in mind for this, is enforced semantic versioning through tooling. This picture shows what we envision for workflow:

<img width="806" alt="screen shot 2015-11-21 at 10 22 38 pm" src="https://cloud.githubusercontent.com/assets/9919222/11321886/765a3f0c-909e-11e5-896f-f5da60ae9464.png">

The commit, release and monitor actions are TODO's. We want to extend the Corto tool with the following commands:

```
corto commit <commit msg>  // Commit code to GitHub
corto publish major  // Increase major version, create tag
corto publish minor  // Increase minor version, create tag
corto publish patch  // Increase patch version, create tag
```

The corto publish commands will trigger builds on Travis which will produce binaries that will be added to the binary repository. They would also update the aforementioned ./.corto/version.txt.

**Comment by SanderMertens** [comment]
Adding support for debug, development & release builds will be done through the corto build command:

```
corto build debug   // Default
corto build dev
corto build release
```

I would like to allow for customizing compiler parameters for these options, but am not quite sure on what level this should be allowed. For example, should a user be able to specify this per project (in the rakefile), in an extern configuration file, the shell environment, or a combination of the above?

**Comment by hendrenja** [comment]
A combination of options would be ideal. I think a robust solution would
include a configuration file that configures the build environment at a
global level. All parameters specified in the configuration file should be
available in the rakefile, or at least items that make sense. Any parameter
specified in the rakefile would override parameters defined in the
configuration file.

A simple way to say that would be, a hierarchy :)

Kind regards,

*Alex Hendren*
*Lead System Architect *
*________________________________________________________________*
Remüv Technologies Inc | 1415 North Loop West, Suite 612,  Houston, TX 77008
 | www.remuv.io
Cell: +1 281.844.0801 | Office: +1 713.485.4519 | Email: *alex.hendren@remuv.io
<alex.hendren@remuv.io>*

*[image: Inline image 1]*

This email and any attachments to it may be confidential and are intended
solely for the use of the individual

to whom it is addressed. Any views or opinions expressed are solely those
of the author and do not

necessarily represent those of Remuv Technologies Inc. If you are not the
intended recipient of this email,

you must neither take any action based upon its contents, nor copy or show
it to anyone. Please contact

the sender if you believe you have received this email in error.

*QPH* *Please consider your environmental responsibility. Save a TREE
! DON'T PRINT this e-mail unless it’s really necessary.*

**Comment by jleeothon** [comment]
Great ideas! It seems to me that it aligns pretty well with the current Corto philosophy of (roughly,) convention-over-configuration: make things dead simple unless you need a funky ad hoc setting.

As a stretch goal, we may want to support build configurations without hard coding the names. So, "dev", "debug" and "release" are not special names except they come by default shipped with Corto. This way, different consumers can create their build configs, e.g. to deploy in "stage" machines \*wink\*.

If agreed, we could think of how to fluently sort-of take priority like: build command => rakefile => local config => global config.

**Comment by hendrenja** [comment]
Hello Johnny, 

Could you please clarify:
>So, "dev", "debug" and "release" are not special names except they come by default shipped with Corto. This way, different consumers can create their build configs, e.g. to deploy in "stage" machines *wink*."

@SanderMertens, will the publish command be the only mechanic to automatically increment the packaging framework?
    corto commit <commit msg>  // Commit code to GitHub
    corto publish major  // Increase major version, create tag
    corto publish minor  // Increase minor version, create tag
    corto publish patch  // Increase patch version, create tag

We discussed defining version in a file like `version.txt` in `$projectDir/.corto`. This would always ensure there is a manual way to define the current version of your package, but it would be nice if there was an intermediate step to the publish. Maybe I am misunderstanding what is happening during publish?

It would be nice to be able to say `corto package major`, which would simply increment our major package label.

**Comment by jleeothon** [comment]
Sure. Uh, can I also ask, what is often the difference between "dev" and "debug" builds? As I would understand it, one is for normal development (usually) on your own machine, and the other one is similar, except it might compile assertions, and log a lot more information.

Where I work, we have "stage" machines, which are environments that are very similar to production, mostly except they don't get traffic from outside, and the databases don't have sensitive information from customers. I'm guessing it could be likely that [a company] that consumes Corto may want to use more than three predefined build environments.

In this way, I'm suggesting that "dev", "debug", and "release" don't need to be hardcoded into the `corto build` logic; instead they could be names of (rakefiles? configuration files?) that come shipped by default with Corto. Nothing should stop you from doing ``corto build environment-A-with-database-X``.

What do you think?

**Comment by SanderMertens** [comment]
@jleeothon I agree dev, debug and release shouldn't be hardcoded, however we can ship supporting these three, and allow users to add more/tweak existing when required.

@hendrenja I see what you're saying. I would like to group all release related activities under a single command ('publish'). What about this:

```
corto publish major --notag   // increases major version, just updates version.txt
corto publish // creates a tag with whatever is in version.txt
```

**Comment by hendrenja** [comment]
@SanderMertens That would be just fine for me!

@jleeothon Okay, I understand your point and I completely agree. I think Corto would solve 90% of use cases with simply debug and release.

**Comment by SanderMertens** [comment]
I implemented the part of the `corto publish` command that manages version.txt. In the next release you'll be able to for example do:
```
$ corto publish MyApp minor --notag
```
The command will throw an error when `--notag` is not provided, as git integration is still a TODO.

**Comment by jleeothon** [comment]
Can you guys update me on version.txt? How does it relate to the version in rakefile and in the source code?

**Comment by SanderMertens** [comment]
There is no version in the rakefile- you probably meant the configure script.

The version you are referring to is the Corto version. version.txt keeps track of a per-project version, so you can do version management on individual packages/applications etc.

**Comment by hendrenja** [comment]
Excellent stuff!

**Comment by hendrenja** [comment]
Dear Trusted Remuv Partner,


As on the 25th May 2017 and with immediate effect, James Alex Hendren, has
resigned from Remuv Technologies Inc.   Mr. Hendren no longer has any
authority to continue discussions with any Remuv employees, partners or
prospects, nor does Mr. Hendren represent Remuv in any way. If you have
entered into any negotiations or agreement with Mr. Hendren, please contact
Remuv  your earliest convenience.    We would like to take this opportunity
to wish Mr. Hendren the very best of luck in his future endeavors.


Kind regards,

Remuv team

**Comment by SanderMertens** [comment]
With `bake` it is now possible to configure different build configurations (`debug`, `dev`, `release`, ...). Version management will be solved by #443.
